File: test/test_native_host.py
import importlib.util
import io
import json
import struct
import unittest
from pathlib import Path


HOST_PATH = Path(__file__).resolve().parents[1] / "native" / "chrome_openwith_mpv.py"


def load_host():
    spec = importlib.util.spec_from_file_location("chrome_openwith_mpv", HOST_PATH)
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module


class NativeHostTests(unittest.TestCase):
    def test_rejects_non_http_urls(self):
        host = load_host()

        result = host.handle_message({"url": "file:///etc/passwd"}, launcher=lambda _url: None)

        self.assertFalse(result["ok"])
        self.assertEqual(result["error"], "unsupported_url")

    def test_launches_http_url(self):
        host = load_host()
        launched = []

        result = host.handle_message(
            {"url": "https://www.twitch.tv/ohnepixel"},
            launcher=lambda url, headers=None: launched.append((url, headers)),
        )

        self.assertEqual(result, {"ok": True})
        self.assertEqual(launched, [("https://www.twitch.tv/ohnepixel", {})])

    def test_launches_http_url_with_headers(self):
        host = load_host()
        launched = []

        result = host.handle_message(
            {
                "url": "https://edgestreams.pro/hls/zeuryuegn48.m3u8?st=abc&e=123",
                "headers": {
                    "Referer": "https://buff-streams.online/UFC250/c48.php",
                    "User-Agent": "Mozilla/5.0",
                },
            },
            launcher=lambda url, headers=None: launched.append((url, headers)),
        )

        self.assertEqual(result, {"ok": True})
        self.assertEqual(
            launched,
            [
                (
                    "https://edgestreams.pro/hls/zeuryuegn48.m3u8?st=abc&e=123",
                    {
                        "Referer": "https://buff-streams.online/UFC250/c48.php",
                        "User-Agent": "Mozilla/5.0",
                    },
                )
            ],
        )

    def test_launches_http_url_with_start_time(self):
        host = load_host()
        launched = []

        result = host.handle_message(
            {"url": "https://www.youtube.com/watch?v=rb3THmr4j2c", "start_time": 42.25},
            launcher=lambda url, headers=None, start_time=None: launched.append((url, headers, start_time)),
        )

        self.assertEqual(result, {"ok": True})
        self.assertEqual(launched, [("https://www.youtube.com/watch?v=rb3THmr4j2c", {}, 42.25)])

    def test_launch_mpv_waits_for_window(self):
        host = load_host()
        calls = []

        class Proc:
            pid = 4321

        def fake_popen(args, **kwargs):
            calls.append(("popen", args, kwargs))
            return Proc()

        def fake_wait(pid, timeout_seconds=15):
            calls.append(("wait", pid, timeout_seconds))
            return True

        host.launch_mpv(
            "https://cdn.example.com/live.m3u8",
            popen=fake_popen,
            window_waiter=fake_wait,
        )

        self.assertEqual(calls[0][0], "popen")
        self.assertEqual(calls[1], ("wait", 4321, 15))

    def test_uses_youtube_profile_for_youtube_urls(self):
        host = load_host()
        config_dir = Path.home() / ".config" / "chrome-openwith-mpv" / "mpv-youtube"

        args = host.build_launch_args("https://www.youtube.com/watch?v=rb3THmr4j2c")

        self.assertIn(f"--config-dir={config_dir}", args)
        self.assertNotIn("--http-header-fields=", " ".join(args))
        self.assertNotIn("cookies-from-browser=brave", " ".join(args))

    def test_build_launch_args_supports_start_time(self):
        host = load_host()

        args = host.build_launch_args(
            "https://cdn.example.com/live.m3u8",
            start_time=123.456,
        )

        self.assertIn("--start=123.456", args)

    def test_reads_and_writes_native_messaging_frames(self):
        host = load_host()
        payload = json.dumps({"url": "https://example.com/watch"}).encode("utf-8")
        stdin = io.BytesIO(struct.pack("<I", len(payload)) + payload)
        stdout = io.BytesIO()
        launched = []

        exit_code = host.run(
            stdin=stdin,
            stdout=stdout,
            launcher=lambda url, headers=None: launched.append((url, headers)),
        )

        self.assertEqual(exit_code, 0)
        self.assertEqual(launched, [("https://example.com/watch", {})])
        stdout.seek(0)
        size = struct.unpack("<I", stdout.read(4))[0]
        response = json.loads(stdout.read(size).decode("utf-8"))
        self.assertEqual(response, {"ok": True})


if __name__ == "__main__":
    unittest.main()

File: test/test_native_host.mjs
import assert from "node:assert/strict";
import { createRequire } from "node:module";
import path from "node:path";
import test from "node:test";

const require = createRequire(import.meta.url);
const HOST_PATH = path.resolve(path.dirname(new URL(import.meta.url).pathname), "../native/chrome_openwith_mpv.js");

function loadHost() {
  delete require.cache[HOST_PATH];
  return require(HOST_PATH);
}

test("rejects non-http urls", () => {
  const host = loadHost();

  const result = host.handleMessage({ url: "file:///etc/passwd" }, () => {});

  assert.equal(result.ok, false);
  assert.equal(result.error, "unsupported_url");
});

test("launches http url", () => {
  const host = loadHost();
  const launched = [];

  const result = host.handleMessage(
    { url: "https://www.twitch.tv/ohnepixel" },
    (url, headers) => launched.push([url, headers])
  );

  assert.deepEqual(result, { ok: true });
  assert.deepEqual(launched, [["https://www.twitch.tv/ohnepixel", {}]]);
});

test("launches http url with headers", () => {
  const host = loadHost();
  const launched = [];

  const result = host.handleMessage(
    {
      url: "https://edgestreams.pro/hls/zeuryuegn48.m3u8?st=abc&e=123",
      headers: {
        Referer: "https://buff-streams.online/UFC250/c48.php",
        "User-Agent": "Mozilla/5.0",
      },
    },
    (url, headers) => launched.push([url, headers])
  );

  assert.deepEqual(result, { ok: true });
  assert.deepEqual(launched, [
    [
      "https://edgestreams.pro/hls/zeuryuegn48.m3u8?st=abc&e=123",
      {
        Referer: "https://buff-streams.online/UFC250/c48.php",
        "User-Agent": "Mozilla/5.0",
      },
    ],
  ]);
});

test("build launch args skips empty headers", () => {
  const host = loadHost();

  const args = host.buildLaunchArgs("https://www.twitch.tv/ohnepixel", {});

  assert.equal(args.some((arg) => arg.startsWith("--http-header-fields=")), false);
});

test("launches http url with start time", () => {
  const host = loadHost();
  const launched = [];

  const result = host.handleMessage(
    { url: "https://www.youtube.com/watch?v=rb3THmr4j2c", start_time: 42.25 },
    (url, headers, startTime) => launched.push([url, headers, startTime])
  );

  assert.deepEqual(result, { ok: true });
  assert.deepEqual(launched, [["https://www.youtube.com/watch?v=rb3THmr4j2c", {}, 42.25]]);
});

test("launch mpv waits for window", () => {
  const host = loadHost();
  const calls = [];

  class Proc {
    constructor() {
      this.pid = 4321;
    }
  }

  host.launchMpv(
    "https://cdn.example.com/live.m3u8",
    null,
    null,
    (command, args, options) => {
      calls.push(["popen", command, args, options]);
      return new Proc();
    },
    (pid, timeoutSeconds) => {
      calls.push(["wait", pid, timeoutSeconds]);
      return true;
    }
  );

  assert.equal(calls[0][0], "popen");
  assert.deepEqual(calls[1], ["wait", 4321, 15]);
});

test("uses youtube profile for youtube urls", () => {
  const host = loadHost();
  const fs = require("node:fs");
  const original = {
    existsSync: fs.existsSync,
    mkdirSync: fs.mkdirSync,
    readFileSync: fs.readFileSync,
    writeFileSync: fs.writeFileSync,
  };

  fs.existsSync = () => false;
  fs.mkdirSync = () => {};
  fs.readFileSync = () => "";
  fs.writeFileSync = () => {};

  try {
    const args = host.buildLaunchArgs("https://www.youtube.com/watch?v=rb3THmr4j2c");

    assert.ok(args.some((arg) => arg.startsWith("--config-dir=")));
    assert.equal(args.some((arg) => arg.startsWith("--http-header-fields=")), false);
  } finally {
    fs.existsSync = original.existsSync;
    fs.mkdirSync = original.mkdirSync;
    fs.readFileSync = original.readFileSync;
    fs.writeFileSync = original.writeFileSync;
  }
});

test("build launch args supports start time", () => {
  const host = loadHost();

  const args = host.buildLaunchArgs("https://cdn.example.com/live.m3u8", null, 123.456);

  assert.ok(args.includes("--start=123.456"));
});

test("reads and writes native messaging frames", () => {
  const host = loadHost();
  const payload = Buffer.from(JSON.stringify({ url: "https://example.com/watch" }), "utf8");
  const length = Buffer.alloc(4);
  length.writeUInt32LE(payload.length, 0);
  const framed = Buffer.concat([length, payload]);
  const writes = [];
  const stdin = {
    cursor: 0,
    read(size) {
      const chunk = framed.subarray(this.cursor, this.cursor + size);
      this.cursor += chunk.length;
      return chunk.length ? chunk : null;
    },
  };
  const stdout = {
    write(chunk) {
      writes.push(Buffer.from(chunk));
    },
  };

  const exitCode = host.run(stdin, stdout, (url, headers) => writes.push(Buffer.from(JSON.stringify([url, headers]))));

  assert.equal(exitCode, 0);
  assert.ok(writes.length > 0);
});
